### ISSUE TITLE
[Feature]: Support for Function Calling or Tools

### ISSUE THREAD

**Issue by dcostea** [SciSharp/LLamaSharp#707]
### Background & Description

Function calling is a feature supported by many open source models, like Mistral.

Function calling feature would allow developers to use LlamaSharp in a more dynamic way, enabling them to perform complex tasks directly within the LlamaSharp environment.

Are there plans to include function calling / tools in LLamaSharp?

**Comment by AsakusaRinne** [comment]
Yes, it is important for using LLM to handling complex works. I'll start making this feature after about 1 month. :)

**Comment by abhiaagarwal** [comment]
It's pretty easy to hack together your own function calling. I used Hermes-2-Theta-LLama-3-8B and got acceptable results. What would be nice is C# native annotations that translate functions into OpenAI JSON, which the "function calling method" meta seems to have standardized around. Semantic Kernel also provides this feature if you really need it with their "planner" functionality, I chose to roll my own for resource constraint reasons

**Comment by AsakusaRinne** [comment]
> It's pretty easy to hack together your own function calling. I used Hermes-2-Theta-LLama-3-8B and got acceptable results. What would be nice is C# native annotations that translate functions into OpenAI JSON, which the "function calling method" meta seems to have standardized around.

That's inspiring! May I ask whether you work is open-source? I wonder if I can take a look of your implementation if it's possible. :)

**Comment by abhiaagarwal** [comment]
Not yet! I'm currently working out kinks, but I hope to get it done ASAP

**Comment by davidlahuta** [comment]
Any progress?

**Comment by Amine-Smahi** [comment]
Any progress?

**Comment by iamadevgh** [comment]
Any progress yet?

**Comment by mkn8rn** [comment]
Any progress?

**Comment by SeriousOldMan** [comment]
I am interested in this as well.

**Comment by mkn8rn** [comment]
WiP: https://github.com/supprocom/LlamaSharp.ToolCallEnvelopes

Edit: link changed

**Comment by SeriousOldMan** [comment]
> WiP: https://github.com/mkn8rn/LlamaSharp.ToolCallEnvelopes

Thx a lot. Will check it out.
